### PR TITLE
Demo: proposed API for @google/events + Functions Framework

### PR DESCRIPTION
**Pull request by matthewrobertson**
This a draft PR that proposes a number of changes to the `@googe/events` library and demonstrates the developer experience provided by the combined `@google/events` and `@google-cloud/functions-framework` APIs

### CODE REVIEW

File: samples/hello_cloudevents/google_events.ts [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  FirebaseAuthV1AuthEventData,
+  FirebaseDatabaseV1ReferenceEventData,
+  FirebaseRemoteconfigV1RemoteConfigEventData,
+  FirebaseTestlabV1TestMatrixEventData,
+};
+
+/**
+ * Modification 3: Declare a mapped type that defines which CloudEvent types correspond
+ * to which data payload schemas. This is an internal implementation detail of the
+ * @google/events library, it is not exposed in the public API.
```
Comment (grayside) [inline samples/hello_cloudevents/google_events.ts:48]: Is this type statement proposed for addition to https://github.com/googleapis/google-cloudevents-nodejs? Initially reacting to the comment, but more generally wondering how many of these type behaviors should we support outside the functions framework?

File: samples/hello_cloudevents/src/multiple_google_events.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import {isGoogleEvent} from '../google_events';
+
+functions.cloudEvent('MultiGoogleEvent1', ce => {
+  if (!isGoogleEvent(ce)) {
+    throw "woops this shouldn't happen!";
+  }
+  // in this branch the typechecker knows ce is a GoogleCloudEvent
+  switch (ce.type) {
+    case 'google.cloud.audit.log.v1.written':
+      ce.data.insertId.length; // OK!
```
[the remark this answers is not in the document]
Comment (jskeet) [reply samples/hello_cloudevents/src/multiple_google_events.ts:16]: Okay, I think I understand how it's not dynamic now... resolving.

[the remark this answers is not in the document]
Comment (jskeet) [reply samples/hello_cloudevents/src/multiple_google_events.ts:16]: (Actually, will double check.)

File: samples/hello_cloudevents/src/multiple_google_events.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ */
+import * as functions from '@google-cloud/functions-framework';
+import {isGoogleEvent} from '../google_events';
+
+functions.cloudEvent('MultiGoogleEvent1', ce => {
+  if (!isGoogleEvent(ce)) {
+    throw "woops this shouldn't happen!";
+  }
+  // in this branch the typechecker knows ce is a GoogleCloudEvent
+  switch (ce.type) {
```
Comment (grant) [inline samples/hello_cloudevents/src/multiple_google_events.ts:14]: I am a little bit concerned about how this function paradigm would scale to less loosely structured languages like Golang.

I think ideally we should strive for declarative type functionality that has the best experience for a single type. Currently, GCF samples only handle single types as well.

File: samples/hello_cloudevents/src/multiple_google_events.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (matthewrobertson) [reply samples/hello_cloudevents/src/multiple_google_events.ts:16]: Typescript is statically type checked. Not all objects have a `length` property, so this line is demonstrating that the type checker knows `ce.data.instertId` is of type `string`

File: samples/hello_cloudevents/google_events.ts (hunk 1 above)
Comment (matthewrobertson) [reply samples/hello_cloudevents/google_events.ts:48]: Yes the proposal is to that all the changes in this file (including this type) would be implemented in the googleapis/google-cloudevents-nodejs library. GCF customers working with cloudevents could choose to take a dependency on that library to get better type checking / autocompletion of their CloudEvent functions.

**Pull request closed without merging**